app/controllers/saml_idp_controller.rb

```diff
@@ -94,27 +94,13 @@ def process_invalid_authn_context
     render nothing: true, status: :bad_request
   end
 
-  def authn_context_node
-    saml_request.document.xpath(
-      '//samlp:AuthnRequest/samlp:RequestedAuthnContext/saml:AuthnContextClassRef',
-      samlp: Saml::XML::Namespaces::PROTOCOL,
-      saml: Saml::XML::Namespaces::ASSERTION)
-  end
-
   def requested_authn_context
-    return authn_context_node[0].content if authn_context_node.length == 1
-
-    logger.info 'authn_context is missing'
-    nil
-  end
-
-  def logout_response_builder
-    SamlIdp::LogoutResponseBuilder.new(
-      get_saml_response_id,
-      issuer_uri,
-      saml_response_url,
-      saml_request.request_id,
-      signature_opts)
+    if saml_request.requested_authn_context
+      saml_request.requested_authn_context
+    else
+      logger.info 'authn_context is missing'
+      nil
+    end
```

Comment:
I have added memoization upstream: https://github.com/pkarman/saml_idp/commit/693ef99be920e4bc7ca57c6a764d1c740e154094

Let's tackle the use of `authn_context` in a separate PR. It seems like there's some app-specific logic that isn't really about the upstream gems, which is the focus of this PR.